### PR TITLE
Automatic hints over time

### PR DESCRIPTION
**Pull request by Alchav**
Adds `minutes_to_hint` to host.yaml which can be set to a number of minutes for all players to gain enough hint points for one hint. Defaults to 0 which disables the option.

Adds a hint_start_time to `/options` which will start at 0. If it is at 0 no hint points will be calculated from it. Can be manually set using `/option hint_start_time {timestamp}` or `/option hint_start_time now` to set it to the current time. However it will automatically set to the current time if a countdown finishes while no checks have yet been made, or if a check is made and hint_start_time is still set to 0. This essentially means it will automatically set to the time that the games began.

### CODE REVIEW

**Comment by LegendaryLinux** [comment]
I feel like if timed hints are enabled, check-based hints should be disabled.

**Comment by Alchav** [comment]
> I feel like if timed hints are enabled, check-based hints should be disabled.

They can be disabled by setting location_check_points to 0. It is up to the host. They can have one, the other, both, or none

**Comment by LegendaryLinux** [comment]
I was just thinking about a situation where someone enables both, and all the players end up with the ability to issue tons of hints. The same could be accomplished by setting the constraint for either option poorly though, so that possibility already exists.

Do remember to also document this new option, please.

File: WebHostLib/templates/generate.html [hunk 1]
```diff
@@ -150,14 +164,69 @@ <h1>Generate Game{% if race %} (Race Mode){% endif %}</h1>
...
+                                                <option value="45">
+                                                    Every 45 minutes
+                                                </option>
+                                                <option value="60">
+                                                    Every 1 hour
+                                                </option>
+                                                <option value="120">
+                                                    Every 2 hours
+                                                </option>
+                                                <option value="12 hours">
```
Comment (Berserker66) [inline WebHostLib/templates/generate.html:217]: The value here looks broken

Comment (Alchav) [reply WebHostLib/templates/generate.html:217]: Good catch. I had all these values and the displayed text swapped initially and had to fix, guess I missed one

File: MultiServer.py [hunk 2]
```diff
@@ -495,10 +505,20 @@ def set_save(self, savedata: dict):
...
+            if self.hint_cost and player in self.locations:
+                self.player_hint_costs[player] = max(0, int(self.hint_cost * 0.01 * len([loc for loc in self.locations[player].values()])))
+            else:
+                self.player_hint_costs[player] = 0
+
     def get_hint_cost(self, slot):
-        if self.hint_cost:
-            return max(0, int(self.hint_cost * 0.01 * len(self.locations[slot])))
-        return 0
+        try:
```
Comment (Berserker66) [inline MultiServer.py:517]: if you run calculate_hint_costs on multidata load and sasvegame load you should never have a KeyError, correct?

File: MultiServer.py [hunk 3]
```diff
@@ -811,6 +834,8 @@ def send_items_to(ctx: Context, team: int, target_slot: int, *items: NetworkItem
 
 def register_location_checks(ctx: Context, team: int, slot: int, locations: typing.Iterable[int],
                              count_activity: bool = True):
+    if ctx.hint_start_time == 0:
+        ctx.hint_start_time = time.time()
```
Comment (Berserker66) [inline MultiServer.py:838]: time.time has platform dependant quirks; it might be better to use datetime.utcnow()

**Comment by Berserker66** [comment]
I think everywhere in server code we use seconds; so "minutes_to_hint" is fine as such in the user-facing space, but internally I'd prefer it be seconds based.

**Comment by Alchav** [comment]
> I think everywhere in server code we use seconds; so "minutes_to_hint" is fine as such in the user-facing space, but internally I'd prefer it be seconds based.

I tried briefly to get this to happen and was getting errors. It seemed needlessly complicated to have to convert numbers and variable names (or having an internal minutes_to_hint variable that does not contain the minutes to hint) at every communication to or from users, so I am giving up on it. I can make it seconds to hint or leave it as is. If someone else wants to implement your idea, they can have at it.

**Pull request closed without merging**